### ISSUE TITLE
`moveTo` may involve copying all the content synchronously

### ISSUE THREAD

**Issue by meejah** [LeastAuthority/magic-folder#425]
`moveTo` may involve copying all the content synchronously. We may want to avoid causing the process to block if we end up copying a multi-GB files here. (We should at least file an issue about this).

See also the comment above about race-conditions when this isn't synchronous.

_Originally posted by @tomprince in https://github.com/LeastAuthority/magic-folder/pull/381#discussion_r651242994_

**Comment by meejah** [comment]
I guess I was assuming this is only "moving" the file on the same filesystem.

Under what circumstances will it be a long-running copy instead of a quick update?

**Comment by tomprince** [comment]
We have no guarantee that the magic-folder and the stash directory are on the same filesystem.

**Comment by meejah** [comment]
If we did guarantee that, would it solve this issue?

**Comment by tomprince** [comment]
> If we did guarantee that, would it solve this issue?

Yes, though if we went that route, we may want to switch to an API that errors if we aren't on the same filesystem.

Doing so would probably involve putting the stash directory inside the magic-folder. Even that isn't necessary enough, as the magic-folder could cross filesystem boundaries. We could treat that as an error though.

I'm not sure which case is more complex, though.

**Comment by meejah** [comment]
Basically the downloader is "doing some async work to 'stage' the file" and then doing some must-be-atomic work to "put it in the right spot".

So either "stash-path must be on the same filesystem" or "something else" so that we can "commit" the file to correct place quickly (the atomic part) after we've downloaded the data (the async part).

Could even using the magic-folder-dir _as_ the stash-path: like use a name like `"._stash<hash-or-something>"` or so before renaming it to `"seekrit.doc"` after we're done with the "async work" part (i.e. downloading).

**Comment by tomprince** [comment]
> Basically the downloader is "doing some async work to 'stage' the file" and then doing some must-be-atomic work to "put it in the right spot".

This sounds right.

> Could even using the magic-folder-dir _as_ the stash-path: like use a name like `"._stash<hash-or-something>"` or so before renaming it to `"seekrit.doc"` after we're done with the "async work" part (i.e. downloading).

Yes. We'd need to ensure that the local snapshotting ignores this file.   `.moveTo` does this under the hood, but we can currently get away with ignoring the temporary file that is created since `.moveTo` is blocking the main thread (i.e. no other code on the main thread should ever be able to observe that temporary file).

**Comment by tomprince** [comment]
Note that the original issue is only relevant for downloading files, and not for the snapshots taken by the uploader.

> Could even using the magic-folder-dir as the stash-path: like use a name like "._stash<hash-or-something>" or so before renaming it to "seekrit.doc" after we're done with the "async work" part (i.e. downloading).

I think this might make the most sense, if we put the temporary file in the same directory as the destination file. If that has issues with cross device moves, If we can't rename in a single directory atomically, I think we'd be forced to overwrite the file in-place, rather than doing an atomic rename.
